scipy/sparse/tests/test_base.py

```diff
@@ -740,6 +740,25 @@ def test_diagonal(self):
         assert_equal(self.spmatrix((15, 0)).diagonal(), np.empty(0))
         assert_equal(self.spmatrix((0, 5)).diagonal(10), np.empty(0))
 
+    def test_trace(self):
+        # For square matrix
+        A = csr_matrix([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
+        desired = A[0, 0] + A[1, 1] + A[2, 2]
+        assert_equal(A.trace(), desired)
+        A = A.tocsc()
+        assert_equal(A.trace(), desired)
+        A = A.tocoo()
+        assert_equal(A.trace(), desired)
+
+        # For rectangular matrix
+        A = csr_matrix([[1, 2, 3], [4, 5, 6]])
+        desired = A[0, 0] + A[1, 1]
```

Comment:
I guess this is the sum of the main diagonal. I got slightly confused because https://en.wikipedia.org/wiki/Trace_(linear_algebra) indicates:

> The trace is only defined for a square matrix (n × n). 

But if I understand correctly we're basically following NumPy convention here in any case.

Reply:
Yes. your understanding is correct. From linear algebra textbook, the standard definition of the trace is only for square matrix. But from algorithm implementation, the definition of the trace can be extended like NumPy but does not conflict with the standard definition (square matrix). In fact, the PR ensure consistency with NumPy convention.